Overlay icon

meandavejustice opened the pull request:
- fixes #64 
- add overlay-icon.js for adding icon to dom and handling actions
- add overlay-icon.css for styling icon
- supports youtube(watch/home) and vimeo(detail/home) as well as vimeo and youtube embeds
- fix orphaned audio streams on `send-to-tab` and `close` fixes #84 

I've tried to find a good percentage ratio for the `width`, `top`, and `left` css attributes for the icon container. Something that will look reasonable for the many different sizes. I'm happy to take suggestions for tweaks to this, but I'd like to not hold up this pr for css polish that can be covered in #72 

### Links for testing:
- [youtube home](https://www.youtube.com)
- [youtube watch](https://www.youtube.com/watch?v=Wjku6BZz3EQ)
- [vimeo home](https://vimeo.com/home)
- [vimeo detail](https://vimeo.com/15540231)
- [vimeo embed](http://www.ubu.com/film/anger_puce.html)

### Testing plan:
1. Install addon
2. test both "youtube watch" and "vimeo detail" links above with the context menu
3. navigate to all links above and make sure videos can be sent to min-vid via overlay icon.

### Known issues:
- launching min-vid from `vimeo detail` or `vimeo embed` also starts the video. It seems `stopImmediatePropagation` and `addEventListener capture argument` aren't working. Will probably file a separate issue to be fixed later.
- There seems to be an issue getting the icons attached when navigating around youtube.com, this seems to be because it is a single page app, and [page-mod](https://duckduckgo.com/l/?kh=-1&uddg=https%3A%2F%2Fdeveloper.mozilla.org%2Fen-US%2Fdocs%2FMozilla%2FAdd-ons%2FSDK%2FHigh-Level_APIs%2Fpage-mod) only runs on new pages. This will probably be a simple fix by just running the attach methods every once in a while. I'm holding out for a bit to see if we can think of a more elegant way than an interval.

6a68 commented:
gah, so many reviews! looking right now

pdehaan commented:
Needs rebase.
Seeing slight issue with the hover background:

1. No hover:
![youtube](https://cloud.githubusercontent.com/assets/557895/17157960/e8769b96-5347-11e6-93e3-07758552799a.png)

2. Hover over video:
![youtube](https://cloud.githubusercontent.com/assets/557895/17157970/ff19fcd0-5347-11e6-8de3-e85215fb6624.png)

3. Hover over icon:
![youtube](https://cloud.githubusercontent.com/assets/557895/17157994/1b4e4b68-5348-11e6-949e-a20026c65d75.png)

pdehaan commented:
Same leak w/ full sized videos on YouTube.com:

![justice_league_official_trailer__2017__ben_affleck_dc_superhero_movie_hd_-_youtube](https://cloud.githubusercontent.com/assets/557895/17158412/0ae62a40-534b-11e6-8073-609cf5d36df7.png)

pdehaan commented:
I only see the overlay on https://vimeo.com/48396889 when I hover over the 100x100 pixels in the top left.

---

No overlays on Vimeo Recommended thumbnails:
![nike_bmx_in_taipei_on_vimeo](https://cloud.githubusercontent.com/assets/557895/17158579/22a863c2-534c-11e6-8dfc-f1e81e23ba79.png)

---

Clicking the overlay in this minified floating video opens min-vid, but toggles video in upper-left. Basically dueling minified PiP players:
![nike_bmx_in_taipei_on_vimeo](https://cloud.githubusercontent.com/assets/557895/17158643/8eb57d52-534c-11e6-817e-7c589798923a.png)

pdehaan commented:
No overlays on Vimeo search results (but I do get a "Send to mini player" context menu item):

![danny_macaskill_in_videos_on_vimeo](https://cloud.githubusercontent.com/assets/557895/17158686/f798e2c8-534c-11e6-97c5-41fa73a259f0.png)

6a68 commented:
minimized view looks kinda funny, and there's no way to re-maximize it, but let's fix in a followup PR:

![](https://www.dropbox.com/s/yjr3ip5hucsxft7/Screen%20Shot%202016-07-26%20at%205.13.16%20PM.png?dl=0&raw=true)

6a68 commented on data/icon-overlay.js:
```diff
@@ -0,0 +1,90 @@
+
+// YOUTUBE HOME PAGE
+// TODO: listen for infinite scrollesque updates and apply this again
+// if necessary
```
File a bug and add the bug number to the TODO

6a68 commented on data/icon-overlay.js:
```diff
@@ -0,0 +1,90 @@
+    const tmp = getTemplate();
+    tmp.addEventListener('click', function(ev) {
+      evNoop(ev);
+      const partialUrl = el.querySelector('.yt-uix-sessionlink').getAttribute('href');
```
`querySelector` returns `null` if nothing's found, so the `.getAttribute` could throw...unless `el` is some special thing

6a68 commented on data/icon-overlay.js:
```diff
@@ -0,0 +1,90 @@
+    tmp.addEventListener('click', function(ev) {
+      evNoop(ev);
+      const partialUrl = el.querySelector('.yt-uix-sessionlink').getAttribute('href');
+      if (partialUrl.length) {
```
Seems weird to check the length of a string. Maybe just `if (partialUrl)`, since it does the same thing?

6a68 commented on data/icon-overlay.js:
```diff
@@ -0,0 +1,90 @@
+          url: 'https://youtube.com' + partialUrl,
+          domain: host
+        });
+      } else console.error('Error parsing url from YT home page', partialUrl); // eslint-disable-line no-console
```
wrap the else clause in curly braces, I've been burned by this ("I'll just add another line to the else clause...")

pdehaan replied:
**PROTIP:** ESLint has a [`curly`](http://eslint.org/docs/rules/curly) for this very thing:

```sh
$ eslint .

/Users/pdehaan/dev/github/mozilla/min-vid/app.js
  135:5  error  Expected { after 'if' condition  curly

/Users/pdehaan/dev/github/mozilla/min-vid/data/resize-listener.js
   7:7  error  Expected { after 'if' condition  curly
  12:5  error  Expected { after 'if' condition  curly

/Users/pdehaan/dev/github/mozilla/min-vid/index.js
   30:5  error  Expected { after 'if' condition  curly
   31:5  error  Expected { after 'else'          curly
   79:7  error  Expected { after 'if' condition  curly
   97:7  error  Expected { after 'if' condition  curly
  113:7  error  Expected { after 'if' condition  curly

✖ 8 problems (8 errors, 0 warnings)
```

---

PR incoming!

6a68 commented on data/icon-overlay.js:
```diff
@@ -0,0 +1,90 @@
+  return containerEl;
+}
+
+function evNoop(ev) {
```
`handleEvent` might be a more readable API name

6a68 replied:
oh wait, I think `handleEvent` actually has special meaning...never mind

6a68 commented on data/icon-overlay.js:
```diff
@@ -0,0 +1,90 @@
+          url: 'https://vimeo.com/' + fauxEl.getAttribute('data-clip-id'),
+          domain: host
+        });
+      } else console.error('Error: failed to locate vimeo url'); // eslint-disable-line no-console
```
else curly braces here too

6a68 commented on index.js:
```diff
@@ -87,54 +86,64 @@ cm.Item({
+  getVimeoUrl(id, function(err, streamUrl) {
+    if (!err) updatePanel({src: streamUrl});
+  });
+}
```
As we add more video providers, it might be nice to just have one function and pass in the stuff that varies as a data object:
`{ url: url, getUrlFn: getVimeoUrl, domain: 'vimeo.com' }`
or
`{ url: url, getUrlFn: getVideoId, domain: 'youtube.com' }`

6a68 commented:
> launching min-vid from vimeo detail or vimeo embed also starts the video. It seems stopImmediatePropagation and addEventListener capture argument aren't working. Will probably file a separate issue to be fixed later.

@meandavejustice Have you tried using a capturing listener, instead of a bubbling listener? Might help, not sure though.

6a68 commented:
@meandavejustice OK, I've done a pass, the code's good overall. Only real concern is not null-checking the `querySelector` call, other than that these are minor stylistic nits. Rebase, fix little things, and merge away :beers:

meandavejustice commented:
* fixed querySelector check cc/ @6a68 
* created `launchVideo` function cc/ @6a68 
* `contextMenuLabel` is now defined in one place
* fixed icon background transparency leak
* created #112 for tracking overlay-icon edge cases cc/ @pdehaan 
* created #111 for adding context menu to vimeo detail page
